### PR TITLE
STORM-821: Adding connection provider interface to decouple jdbc conn…

### PR DESCRIPTION
**Pull request by Parth-Brahmbhatt**
…ector from a single connection pooling implementation.

### CODE REVIEW

**Comment by harshach** [comment]
+1

**Comment by ptgoetz** [comment]
+1

File: external/storm-jdbc/README.md [hunk 1]
```diff
@@ -1,10 +1,37 @@
...
 Storm/Trident integration for JDBC. This package includes the core bolts and trident states that allows a storm topology
 to either insert storm tuples in a database table or to execute select queries against a database and enrich tuples 
-in a storm topology. This code uses HikariCP for connection pooling. See http://brettwooldridge.github.io/HikariCP.
+in a storm topology.
 
 ## Inserting into a database.
 The bolt and trident state included in this package for inserting data into a database tables are tied to a single table.
+
+### ConnectionProvider
+An interface that should be implemented by different connection pooling mechanism `org.apache.storm.jdbc.common.ConnectionPrvoider`
```
Comment (HeartSaVioR) [inline external/storm-jdbc/README.md]: There's a typo, ConnectionProvioder -> ConnectionProvider.

**Comment by HeartSaVioR** [comment]
Class name has a typo: ConnectionPrvoider -> ConnectionProvider.
It's not just documentation typo, but also real classname typo.

File: external/storm-jdbc/README.md [hunk 2]
```diff
@@ -17,11 +44,12 @@ The `getColumns()` method defines how a storm tuple maps to a list of columns re
...
 of the returned list of `getColumns` method will map to the 1st place holder and the 2nd to the 2nd and so on. We do not parse
-the supplied queries to try and resolve place holder by column names. 
+the supplied queries to try and resolve place holder by column names. Not making any assumptions about the query syntax allows this connector
+to be used by some non-standard sql frameworks like Pheonix which only supports upsert into.
 
 ### JdbcInsertBolt
-To use the `JdbcInsertBolt`, you construct an instance of it and specify a configuration key in your storm config that holds the 
-hikari configuration map and a `JdbcMapper` implementation that coverts storm tuple to DB row. In addition, you must either supply 
+To use the `JdbcInsertBolt`, you construct an instance of it by specifying a `ConnectionProvider` implementation
+and a `JdbcMapper` implementation that coverts storm tuple to DB row. In addition, you must either supply
```
Comment (HeartSaVioR) [inline external/storm-jdbc/README.md]: coverts -> converts

File: external/storm-jdbc/README.md [hunk 3]
```diff
@@ -29,13 +57,21 @@ The default is set to value of topology.message.timeout.secs and a value of -1 w
 You should set the query timeout value to be <= topology.message.timeout.secs.
 
  ```java
-Config config = new Config();
-config.put("jdbc.conf", hikariConfigMap);
-JdbcInsertBolt userPersistanceBolt = new JdbcInsertBolt("jdbc.conf",simpleJdbcMapper)
+Map hikariConfigMap = Maps.newHashMap();
```
Comment (HeartSaVioR) [inline external/storm-jdbc/README.md]: minor thing: 
How about changing this line to 
```
Map<String, String> hikariConfigMap = new HashMap<String, String>();
```
to remove dependency about Guava from example?

File: external/storm-jdbc/src/main/java/org/apache/storm/jdbc/mapper/SimpleJdbcMapper.java [hunk 4]
```diff
@@ -33,9 +34,10 @@
 
     private List<Column> schemaColumns;
 
-    public SimpleJdbcMapper(String tableName, Map hikariConfigurationMap) {
+    public SimpleJdbcMapper(String tableName, Map<String, Object> connectionProvideConfig, ConnectionProvider connectionProvider) {
```
Comment (HeartSaVioR) [inline external/storm-jdbc/src/main/java/org/apache/storm/jdbc/mapper/SimpleJdbcMapper.java]: Seems like we don't need to connectionProvideConfig here for now.

File: external/storm-jdbc/README.md [hunk 5]
```diff
@@ -60,6 +96,7 @@ hikariConfigMap.put("dataSourceClassName","com.mysql.jdbc.jdbc2.optional.MysqlDa
 hikariConfigMap.put("dataSource.url", "jdbc:mysql://localhost/test");
 hikariConfigMap.put("dataSource.user","root");
 hikariConfigMap.put("dataSource.password","password");
+ConnectionProvider connectionProvider = new HikariCPConnectionProvider(map);
 String tableName = "user_details";
 JdbcMapper simpleJdbcMapper = new SimpleJdbcMapper(tableName, map);
```
Comment (HeartSaVioR) [inline external/storm-jdbc/README.md]: After modifying https://github.com/apache/storm/pull/556/files#r31087884, it should be changed to 
```
JdbcMapper simpleJdbcMapper = new SimpleJdbcMapper(tableName, connectionProvider);
```

**Comment by Parth-Brahmbhatt** [comment]
@HeartSaVioR  all comments addressed, thanks for finding the README issues and keeping the documentation honest. I did not remove the Guava dependency as there are multiple references to Lists and I think Guava is so widely used its unlikely to confuse anyone reading the README.

**Comment by Parth-Brahmbhatt** [comment]
@harshach @ptgoetz @HeartSaVioR I forgot to mention that I also made a minor change in column.java so we can handle database columns with null values. It was a small change so filing another jira felt like an overkill. If anyone has objection to that let me know and I will open a different jira.

**Comment by HeartSaVioR** [comment]
@Parth-Brahmbhatt I agree that Guava is widely-used library. I'm OK to keep dependency from example. :)

File: external/storm-jdbc/README.md (hunk 5 above)
Comment (HeartSaVioR) [inline external/storm-jdbc/README.md]: @Parth-Brahmbhatt This line is not changed yet, so you may want to check it again.

**Comment by HeartSaVioR** [comment]
LGTM. :+1: 

Btw, we don't release storm-jdbc module to any Storm releases yet, so IMO it could be not mandatory to have a new issue regarding to Column when we promise #556 will be merged to 0.10.x before releasing.

**Comment by ptgoetz** [comment]
Thanks @Parth-Brahmbhatt. I merged this into master and 0.10.x.

**Pull request merged**